File: java/com/google/copybara/git/gerritapi/ChangesQuery.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git.gerritapi;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableSet;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import javax.annotation.Nullable;
import net.starlark.java.annot.StarlarkBuiltin;
import net.starlark.java.eval.Printer;
import net.starlark.java.eval.StarlarkValue;

/**
 * An object that represents the input parameters for a changes query:
 *
 * <p>https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes
 */
@StarlarkBuiltin(
    name = "gerritapi.ChangesQuery",
    doc =
        "Input for listing Gerrit changes. See "
            + "https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes")
public class ChangesQuery implements StarlarkValue {

  private final String query;
  private final ImmutableSet<IncludeResult> include;
  @Nullable private final Integer limit;
  @Nullable private final Integer start;

  public ChangesQuery(String query) {
    this.query = query;
    include = ImmutableSet.of();
    limit = null;
    start = null;
  }

  private ChangesQuery(String query, ImmutableSet<IncludeResult> include, Integer limit,
      Integer start) {
    this.query = Preconditions.checkNotNull(query);
    this.include = Preconditions.checkNotNull(include);
    this.limit = limit;
    this.start = start;
  }

  public ChangesQuery withStart(int start) {
    return new ChangesQuery(query, include, limit, start);
  }

  public ChangesQuery withLimit(int limit) {
    return new ChangesQuery(query, include, limit, start);
  }

  public ChangesQuery withInclude(Iterable<IncludeResult> include) {
    return new ChangesQuery(query, ImmutableSet.copyOf(include), limit, start);
  }

  @VisibleForTesting
  public String asUrlParams() {
    StringBuilder sb = new StringBuilder("q=").append(escape(query));
    for (IncludeResult includeResult : include) {
      sb.append("&o=").append(includeResult);
    }
    if (limit != null) {
      sb.append("&n=").append(limit);
    }
    if (start != null) {
      sb.append("&S=").append(start);
    }
    return sb.toString();
  }

  private static String escape(String query) {
    try {
      return URLEncoder.encode(query, "UTF-8");
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException("Shouldn't fail", e);
    }
  }

  @Override
  public void repr(Printer printer) {
    printer.append(toString());
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("query", query)
        .add("include", include)
        .add("limit", limit)
        .add("start", start)
        .toString();
  }
}

File: src/Copybara.Core/Git/GerritApi/ChangesQuery.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Globalization;
using System.Text;
using Copybara.Common;
using Starlark.Eval;

namespace Copybara.Git.GerritApi;

/// <summary>
/// An object that represents the input parameters for a changes query:
///
/// <para>https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes</para>
/// </summary>
[Starlark.Annot.StarlarkBuiltin(
    "gerritapi.ChangesQuery",
    Doc =
        "Input for listing Gerrit changes. See "
        + "https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes")]
public class ChangesQuery : IStarlarkPrintableValue
{
    private readonly string _query;
    private readonly IReadOnlySet<IncludeResult> _include;
    private readonly int? _limit;
    private readonly int? _start;

    public ChangesQuery(string query)
    {
        _query = query;
        _include = ImmutableHashSet<IncludeResult>.Empty;
        _limit = null;
        _start = null;
    }

    private ChangesQuery(string query, IReadOnlySet<IncludeResult> include, int? limit, int? start)
    {
        _query = Preconditions.CheckNotNull(query);
        _include = Preconditions.CheckNotNull(include);
        _limit = limit;
        _start = start;
    }

    public ChangesQuery WithStart(int start) => new(_query, _include, _limit, start);

    public ChangesQuery WithLimit(int limit) => new(_query, _include, limit, _start);

    public ChangesQuery WithInclude(IEnumerable<IncludeResult> include) =>
        new(_query, include.ToImmutableHashSet(), _limit, _start);

    public string AsUrlParams()
    {
        var sb = new StringBuilder("q=").Append(Escape(_query));
        foreach (var includeResult in _include)
        {
            sb.Append("&o=").Append(includeResult);
        }

        if (_limit != null)
        {
            sb.Append("&n=").Append(_limit.Value.ToString(CultureInfo.InvariantCulture));
        }

        if (_start != null)
        {
            sb.Append("&S=").Append(_start.Value.ToString(CultureInfo.InvariantCulture));
        }

        return sb.ToString();
    }

    private static string Escape(string query) => Uri.EscapeDataString(query);

    public void Repr(Printer printer, StarlarkSemantics semantics) => printer.Append(ToString());

    public override string ToString() =>
        $"ChangesQuery{{query={_query}, include=[{string.Join(", ", _include)}], "
        + $"limit={_limit?.ToString(CultureInfo.InvariantCulture) ?? "null"}, "
        + $"start={_start?.ToString(CultureInfo.InvariantCulture) ?? "null"}}}";
}
